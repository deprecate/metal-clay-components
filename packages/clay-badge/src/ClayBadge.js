'use strict';

import 'clay-icon';
import 'clay-link';
import Component from 'metal-component';
import Soy from 'metal-soy';
import { Config } from 'metal-state';

import templates from './ClayBadge.soy.js';

/**
 * Metal Clay Badge component.
 */
class ClayBadge extends Component {}

/**
 * State definition.
 * @static
 * #type {!Object}
 */
ClayBadge.STATE = {
	/**
	 * True or false to activate the close button.
	 * @instance
	 * @memberof ClayBadge
	 * @type {?bool|undefined}
	 * @default undefined
	 */
	closeable: Config.bool(),

	/**
	 * HREF in all badge.
	 * @instance
	 * @memberof ClayBadge
	 * @type {?string|undefined}
	 * default undefined
	 */
	href: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayBadge
	 * @type {?string|undefined}
	 * default undefined
	 */
	id: Config.string(),

	/**
	 * The label of the badge element.
	 * @instance
	 * @memberof ClayBadge
	 * @type {?string|html|undefined}
	 * default undefined
	 */
	label: Config.any(),

	/**
	 * The size of the badge element.
	 * @instance
	 * @memberof ClayBadge
	 * @type {?string|undefined}
	 * default undefined
	 */
	size: Config.oneOf(['sm', 'lg']),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayBadge
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string(),

	/**
	 * Badge style color.
	 * @instance
	 * @memberof ClayBadge
	 * @type {?string|undefined}
	 * default undefined
	 */
	style: Config.oneOf([
		'primary',
		'secondary',
		'success',
		'info',
		'warning',
		'danger',
		'light',
		'dark',
	]).value('primary'),
};

Soy.register(ClayBadge, templates);

export { ClayBadge };
export default ClayBadge;
