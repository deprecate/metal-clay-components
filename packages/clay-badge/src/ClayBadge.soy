{namespace ClayBadge}

/**
 * This renders a badge.
 */
{template .render}
	{@param? closeable: bool}
	{@param? href: string}
	{@param? id: string}
	{@param? label: html|string}
	{@param? size: string}
	{@param? spritemap: string}
	{@param? style: string}

	{let $classes kind="text"}
		badge badge-pill
		{sp}badge-{$style ?: 'default'}

		{if $size}
			{sp}badge-{$size}
		{/if}
	{/let}

	{let $attributes kind="attributes"}
		class="{$classes}"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	{if $href}
		{call ClayLink.render}
			{param elementClasses: $classes /}
			{param href: $href /}
			{param id: $id /}
			{param label: $label /}
		{/call}
	{else}
		<span {$attributes}>
			{call .content}
				{param closeable: $closeable /}
				{param label: $label /}
				{param spritemap: $spritemap /}
			{/call}
		</span>
	{/if}

{/template}

/**
 * Renders icon content
 */
{template .content}
	{@param? closeable: bool}
	{@param? label: html|string}
	{@param? spritemap: string}

	{$label ? $label : ''}

	{if $closeable and $spritemap}
		<button aria-label="Close" class="close" type="button">
			{call ClayIcon.render}
				{param spritemap: $spritemap /}
				{param symbol: 'times' /}
			{/call}
		</button>
	{/if}
{/template}