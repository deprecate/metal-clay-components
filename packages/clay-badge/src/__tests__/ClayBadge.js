'use strict';

import ClayBadge from '../ClayBadge';

let badge;

describe('ClayBadge', function() {
	afterEach(() => {
		if (badge) {
			badge.dispose();
		}
	});

	it('should generate the default markup', () => {
		badge = new ClayBadge();

		expect(badge).toMatchSnapshot();
	});

	it('should render a badge with label', () => {
		badge = new ClayBadge({
			label: 'Primary',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a badge with id', () => {
		badge = new ClayBadge({
			id: 'myBadge',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a badge with href', () => {
		badge = new ClayBadge({
			href: '#1',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a small badge', () => {
		badge = new ClayBadge({
			size: 'sm',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a large badge', () => {
		badge = new ClayBadge({
			size: 'lg',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a badge with closeable', () => {
		badge = new ClayBadge({
			label: 'Badge',
			closeable: {
				monospaced: false,
				spritemap: 'icons.svg',
				symbol: 'times',
			},
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a `primary color` badge', () => {
		badge = new ClayBadge({
			style: 'primary',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a `secondary color` badge', () => {
		badge = new ClayBadge({
			style: 'secondary',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a `success color` badge', () => {
		badge = new ClayBadge({
			style: 'success',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a `info color` badge', () => {
		badge = new ClayBadge({
			style: 'info',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a `warning color` badge', () => {
		badge = new ClayBadge({
			style: 'warning',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a `danger color` badge', () => {
		badge = new ClayBadge({
			style: 'danger',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a `light color` badge', () => {
		badge = new ClayBadge({
			style: 'light',
		});

		expect(badge).toMatchSnapshot();
	});

	it('should render a `dark color` badge', () => {
		badge = new ClayBadge({
			style: 'dark',
		});

		expect(badge).toMatchSnapshot();
	});
});
