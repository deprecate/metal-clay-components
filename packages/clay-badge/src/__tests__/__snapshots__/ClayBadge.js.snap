// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayBadge should generate the default markup 1`] = `<span class="badge badge-pill badge-primary"></span>`;

exports[`ClayBadge should render a \`danger color\` badge 1`] = `<span class="badge badge-pill badge-danger"></span>`;

exports[`ClayBadge should render a \`dark color\` badge 1`] = `<span class="badge badge-pill badge-dark"></span>`;

exports[`ClayBadge should render a \`default color\` badge 1`] = `<span class="badge badge-pill badge-primary"></span>`;

exports[`ClayBadge should render a \`info color\` badge 1`] = `<span class="badge badge-pill badge-info"></span>`;

exports[`ClayBadge should render a \`light color\` badge 1`] = `<span class="badge badge-pill badge-light"></span>`;

exports[`ClayBadge should render a \`primary color\` badge 1`] = `<span class="badge badge-pill badge-primary"></span>`;

exports[`ClayBadge should render a \`secondary color\` badge 1`] = `<span class="badge badge-pill badge-secondary"></span>`;

exports[`ClayBadge should render a \`success color\` badge 1`] = `<span class="badge badge-pill badge-success"></span>`;

exports[`ClayBadge should render a \`warning color\` badge 1`] = `<span class="badge badge-pill badge-warning"></span>`;

exports[`ClayBadge should render a badge with closeable 1`] = `
<span class="badge badge-pill badge-primary">Badge
  <button aria-label="Close" class="close" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
      <use xlink:href="icons.svg#times"></use>
    </svg>
  </button>
</span>
`;

exports[`ClayBadge should render a badge with href 1`] = `<a class=" badge badge-pill badge-primary" href="#1"></a>`;

exports[`ClayBadge should render a badge with id 1`] = `<span class="badge badge-pill badge-primary" id="myBadge"></span>`;

exports[`ClayBadge should render a badge with label 1`] = `<span class="badge badge-pill badge-primary">Primary</span>`;

exports[`ClayBadge should render a large badge 1`] = `<span class="badge badge-pill badge-primary badge-lg"></span>`;

exports[`ClayBadge should render a small badge 1`] = `<span class="badge badge-pill badge-primary badge-sm"></span>`;
